File: playmidifile/android/src/main/kotlin/com/example/playmidifile/PlayMidifilePlugin.kt
package com.example.playmidifile

import android.content.Context
import android.media.MediaPlayer
import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.util.*

/** PlayMidifilePlugin */
class PlayMidifilePlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private lateinit var progressChannel: EventChannel
  private lateinit var stateChannel: EventChannel
  private var context: Context? = null
  
  private var mediaPlayer: MediaPlayer? = null
  private var progressEventSink: EventChannel.EventSink? = null
  private var stateEventSink: EventChannel.EventSink? = null
  private val handler = Handler(Looper.getMainLooper())
  private var progressTimer: Timer? = null
  
  private var currentState = "stopped"
  private var duration = 0
  private var currentPosition = 0

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "playmidifile")
    channel.setMethodCallHandler(this)
    
    progressChannel = EventChannel(flutterPluginBinding.binaryMessenger, "playmidifile/progress")
    progressChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        progressEventSink = events
      }
      
      override fun onCancel(arguments: Any?) {
        progressEventSink = null
      }
    })
    
    stateChannel = EventChannel(flutterPluginBinding.binaryMessenger, "playmidifile/state")
    stateChannel.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        stateEventSink = events
      }
      
      override fun onCancel(arguments: Any?) {
        stateEventSink = null
      }
    })
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initialize" -> {
        initialize(result)
      }
      "loadFile" -> {
        val filePath = call.argument<String>("filePath")
        if (filePath != null) {
          loadFile(filePath, result)
        } else {
          result.error("INVALID_ARGUMENT", "文件路径不能为空", null)
        }
      }
      "loadAsset" -> {
        val assetPath = call.argument<String>("assetPath")
        if (assetPath != null) {
          loadAsset(assetPath, result)
        } else {
          result.error("INVALID_ARGUMENT", "资源路径不能为空", null)
        }
      }
      "play" -> {
        play(result)
      }
      "pause" -> {
        pause(result)
      }
      "stop" -> {
        stop(result)
      }
      "seekTo" -> {
        val positionMs = call.argument<Int>("positionMs")
        if (positionMs != null) {
          seekTo(positionMs, result)
        } else {
          result.error("INVALID_ARGUMENT", "位置参数不能为空", null)
        }
      }
      "setSpeed" -> {
        val speed = call.argument<Double>("speed")
        if (speed != null) {
          setSpeed(speed.toFloat(), result)
        } else {
          result.error("INVALID_ARGUMENT", "速度参数不能为空", null)
        }
      }
      "setVolume" -> {
        val volume = call.argument<Double>("volume")
        if (volume != null) {
          setVolume(volume.toFloat(), result)
        } else {
          result.error("INVALID_ARGUMENT", "音量参数不能为空", null)
        }
      }
      "getCurrentState" -> {
        result.success(currentState)
      }
      "getCurrentInfo" -> {
        getCurrentInfo(result)
      }
      "dispose" -> {
        dispose(result)
      }
      else -> {
        result.notImplemented()
      }
    }
  }
  
  private fun initialize(result: Result) {
    try {
      result.success(null)
    } catch (e: Exception) {
      result.error("INIT_ERROR", "初始化失败: ${e.message}", null)
    }
  }
  
  private fun loadFile(filePath: String, result: Result) {
    try {
      releaseMediaPlayer()
      
      val file = File(filePath)
      if (!file.exists()) {
        result.error("FILE_NOT_FOUND", "文件不存在: $filePath", null)
        return
      }
      
      mediaPlayer = MediaPlayer().apply {
        setDataSource(filePath)
        prepareAsync()
        setOnPreparedListener { mp ->
          this@PlayMidifilePlugin.duration = mp.duration
          updateState("stopped")
          result.success(true)
        }
        setOnErrorListener { _, what, extra ->
          updateState("error")
          result.error("LOAD_ERROR", "加载文件失败: what=$what, extra=$extra", null)
          true
        }
        setOnCompletionListener {
          updateState("stopped")
          stopProgressTimer()
        }
      }
    } catch (e: Exception) {
      result.error("LOAD_ERROR", "加载文件失败: ${e.message}", null)
    }
  }
  
  private fun loadAsset(assetPath: String, result: Result) {
    try {
      releaseMediaPlayer()
      
      val flutterLoader = io.flutter.FlutterInjector.instance().flutterLoader()
      val assetKey = flutterLoader.getLookupKeyForAsset(assetPath)
      
      val assetManager = context?.assets
      if (assetManager == null) {
        result.error("CONTEXT_ERROR", "无法获取应用上下文", null)
        return
      }
      
      mediaPlayer = MediaPlayer().apply {
        try {
          val afd = assetManager.openFd(assetKey)
          setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
          afd.close()
        } catch (e: Exception) {
          // 如果通过Flutter asset路径失败，尝试直接使用传入的路径
          val afd = assetManager.openFd(assetPath)
          setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
          afd.close()
        }
        
        prepareAsync()
        setOnPreparedListener { mp ->
          this@PlayMidifilePlugin.duration = mp.duration
          updateState("stopped")
          result.success(true)
        }
        setOnErrorListener { _, what, extra ->
          updateState("error")
          result.error("LOAD_ERROR", "加载资源失败: $assetPath, what=$what, extra=$extra", null)
          true
        }
        setOnCompletionListener {
          updateState("stopped")
          stopProgressTimer()
        }
      }
    } catch (e: Exception) {
      result.error("LOAD_ERROR", "加载资源失败: ${e.message}", null)
    }
  }
  
  private fun play(result: Result) {
    try {
      val mp = mediaPlayer
      if (mp == null) {
        result.error("NO_FILE", "请先加载文件", null)
        return
      }
      
      mp.start()
      updateState("playing")
      startProgressTimer()
      result.success(null)
    } catch (e: Exception) {
      result.error("PLAY_ERROR", "播放失败: ${e.message}", null)
    }
  }
  
  private fun pause(result: Result) {
    try {
      val mp = mediaPlayer
      if (mp == null) {
        result.error("NO_FILE", "请先加载文件", null)
        return
      }
      
      if (mp.isPlaying) {
        mp.pause()
        updateState("paused")
        stopProgressTimer()
      }
      result.success(null)
    } catch (e: Exception) {
      result.error("PAUSE_ERROR", "暂停失败: ${e.message}", null)
    }
  }
  
  private fun stop(result: Result) {
    try {
      val mp = mediaPlayer
      if (mp != null) {
        if (mp.isPlaying) {
          mp.stop()
        }
        mp.prepareAsync()
        updateState("stopped")
        stopProgressTimer()
      }
      result.success(null)
    } catch (e: Exception) {
      result.error("STOP_ERROR", "停止失败: ${e.message}", null)
    }
  }
  
  private fun seekTo(positionMs: Int, result: Result) {
    try {
      val mp = mediaPlayer
      if (mp == null) {
        result.error("NO_FILE", "请先加载文件", null)
        return
      }
      
      mp.seekTo(positionMs)
      currentPosition = positionMs
      updateProgress()
      result.success(null)
    } catch (e: Exception) {
      result.error("SEEK_ERROR", "跳转失败: ${e.message}", null)
    }
  }
  
  private fun setSpeed(speed: Float, result: Result) {
    try {
      val mp = mediaPlayer
      if (mp == null) {
        result.error("NO_FILE", "请先加载文件", null)
        return
      }
      
      if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
        val params = mp.playbackParams
        params.speed = speed
        mp.playbackParams = params
      }
      result.success(null)
    } catch (e: Exception) {
      result.error("SPEED_ERROR", "设置播放速度失败: ${e.message}", null)
    }
  }
  
  private fun setVolume(volume: Float, result: Result) {
    try {
      val mp = mediaPlayer
      if (mp == null) {
        result.error("NO_FILE", "请先加载文件", null)
        return
      }
      
      mp.setVolume(volume, volume)
      result.success(null)
    } catch (e: Exception) {
      result.error("VOLUME_ERROR", "设置音量失败: ${e.message}", null)
    }
  }
  
  private fun getCurrentInfo(result: Result) {
    try {
      val mp = mediaPlayer
      if (mp != null) {
        currentPosition = mp.currentPosition
        val progress = if (duration > 0) currentPosition.toDouble() / duration.toDouble() else 0.0
        
        val info = mapOf(
          "currentPositionMs" to currentPosition,
          "durationMs" to duration,
          "progress" to progress
        )
        result.success(info)
      } else {
        result.success(null)
      }
    } catch (e: Exception) {
      result.error("INFO_ERROR", "获取播放信息失败: ${e.message}", null)
    }
  }
  
  private fun dispose(result: Result) {
    try {
      releaseMediaPlayer()
      stopProgressTimer()
      result.success(null)
    } catch (e: Exception) {
      result.error("DISPOSE_ERROR", "释放资源失败: ${e.message}", null)
    }
  }
  
  private fun releaseMediaPlayer() {
    mediaPlayer?.let { mp ->
      try {
        if (mp.isPlaying) {
          mp.stop()
        }
        mp.release()
      } catch (e: Exception) {
        // 忽略释放时的错误
      }
    }
    mediaPlayer = null
  }
  
  private fun updateState(newState: String) {
    currentState = newState
    handler.post {
      stateEventSink?.success(newState)
    }
  }
  
  private fun startProgressTimer() {
    stopProgressTimer()
    progressTimer = Timer()
    progressTimer?.scheduleAtFixedRate(object : TimerTask() {
      override fun run() {
        updateProgress()
      }
    }, 0, 100) // 每100ms更新一次
  }
  
  private fun stopProgressTimer() {
    progressTimer?.cancel()
    progressTimer = null
  }
  
  private fun updateProgress() {
    try {
      val mp = mediaPlayer
      if (mp != null && currentState == "playing") {
        currentPosition = mp.currentPosition
        val progress = if (duration > 0) currentPosition.toDouble() / duration.toDouble() else 0.0
        
        val info = mapOf(
          "currentPositionMs" to currentPosition,
          "durationMs" to duration,
          "progress" to progress
        )
        
        handler.post {
          progressEventSink?.success(info)
        }
      }
    } catch (e: Exception) {
      // 忽略更新进度时的错误
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    releaseMediaPlayer()
    stopProgressTimer()
  }
} 

File: playmidifile/ios/Classes/PlayMidifilePlugin.swift
import Flutter
import UIKit
import AVFoundation
import MediaPlayer

public class PlayMidifilePlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var progressEventChannel: FlutterEventChannel?
    private var stateEventChannel: FlutterEventChannel?
    private var progressEventSink: FlutterEventSink?
    private var stateEventSink: FlutterEventSink?
    
    private var audioPlayer: AVAudioPlayer?
    private var progressTimer: Timer?
    private var currentState = "stopped"
    private var duration: TimeInterval = 0
    private var currentPosition: TimeInterval = 0
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "playmidifile", binaryMessenger: registrar.messenger())
        let progressChannel = FlutterEventChannel(name: "playmidifile/progress", binaryMessenger: registrar.messenger())
        let stateChannel = FlutterEventChannel(name: "playmidifile/state", binaryMessenger: registrar.messenger())
        
        let instance = PlayMidifilePlugin()
        instance.methodChannel = channel
        instance.progressEventChannel = progressChannel
        instance.stateEventChannel = stateChannel
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        progressChannel.setStreamHandler(instance)
        stateChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(result: result)
        case "loadFile":
            guard let args = call.arguments as? [String: Any],
                  let filePath = args["filePath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "文件路径不能为空", details: nil))
                return
            }
            loadFile(filePath: filePath, result: result)
        case "loadAsset":
            guard let args = call.arguments as? [String: Any],
                  let assetPath = args["assetPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "资源路径不能为空", details: nil))
                return
            }
            loadAsset(assetPath: assetPath, result: result)
        case "play":
            play(result: result)
        case "pause":
            pause(result: result)
        case "stop":
            stop(result: result)
        case "seekTo":
            guard let args = call.arguments as? [String: Any],
                  let positionMs = args["positionMs"] as? Int else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "位置参数不能为空", details: nil))
                return
            }
            seekTo(positionMs: positionMs, result: result)
        case "setSpeed":
            guard let args = call.arguments as? [String: Any],
                  let speed = args["speed"] as? Double else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "速度参数不能为空", details: nil))
                return
            }
            setSpeed(speed: Float(speed), result: result)
        case "setVolume":
            guard let args = call.arguments as? [String: Any],
                  let volume = args["volume"] as? Double else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "音量参数不能为空", details: nil))
                return
            }
            setVolume(volume: Float(volume), result: result)
        case "getCurrentState":
            result(currentState)
        case "getCurrentInfo":
            getCurrentInfo(result: result)
        case "dispose":
            dispose(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(result: @escaping FlutterResult) {
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)
            result(nil)
        } catch {
            result(FlutterError(code: "INIT_ERROR", message: "初始化失败: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func loadFile(filePath: String, result: @escaping FlutterResult) {
        do {
            releaseAudioPlayer()
            
            let url = URL(fileURLWithPath: filePath)
            guard FileManager.default.fileExists(atPath: filePath) else {
                result(FlutterError(code: "FILE_NOT_FOUND", message: "文件不存在: \(filePath)", details: nil))
                return
            }
            
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.delegate = self
            audioPlayer?.prepareToPlay()
            
            duration = audioPlayer?.duration ?? 0
            updateState("stopped")
            result(true)
        } catch {
            result(FlutterError(code: "LOAD_ERROR", message: "加载文件失败: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func loadAsset(assetPath: String, result: @escaping FlutterResult) {
        do {
            releaseAudioPlayer()
            
            guard let path = Bundle.main.path(forResource: assetPath, ofType: nil) else {
                result(FlutterError(code: "FILE_NOT_FOUND", message: "资源文件不存在: \(assetPath)", details: nil))
                return
            }
            
            let url = URL(fileURLWithPath: path)
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.delegate = self
            audioPlayer?.prepareToPlay()
            
            duration = audioPlayer?.duration ?? 0
            updateState("stopped")
            result(true)
        } catch {
            result(FlutterError(code: "LOAD_ERROR", message: "加载资源失败: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func play(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        player.play()
        updateState("playing")
        startProgressTimer()
        result(nil)
    }
    
    private func pause(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        if player.isPlaying {
            player.pause()
            updateState("paused")
            stopProgressTimer()
        }
        result(nil)
    }
    
    private func stop(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        player.stop()
        player.currentTime = 0
        updateState("stopped")
        stopProgressTimer()
        result(nil)
    }
    
    private func seekTo(positionMs: Int, result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        let timeInSeconds = Double(positionMs) / 1000.0
        player.currentTime = timeInSeconds
        currentPosition = timeInSeconds
        updateProgress()
        result(nil)
    }
    
    private func setSpeed(speed: Float, result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        if #available(iOS 10.0, *) {
            player.rate = speed
        }
        result(nil)
    }
    
    private func setVolume(volume: Float, result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NO_FILE", message: "请先加载文件", details: nil))
            return
        }
        
        player.volume = volume
        result(nil)
    }
    
    private func getCurrentInfo(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(nil)
            return
        }
        
        currentPosition = player.currentTime
        let progress = duration > 0 ? currentPosition / duration : 0.0
        
        let info: [String: Any] = [
            "currentPositionMs": Int(currentPosition * 1000),
            "durationMs": Int(duration * 1000),
            "progress": progress
        ]
        result(info)
    }
    
    private func dispose(result: @escaping FlutterResult) {
        releaseAudioPlayer()
        stopProgressTimer()
        result(nil)
    }
    
    private func releaseAudioPlayer() {
        audioPlayer?.stop()
        audioPlayer = nil
    }
    
    private func updateState(_ newState: String) {
        currentState = newState
        DispatchQueue.main.async {
            self.stateEventSink?(newState)
        }
    }
    
    private func startProgressTimer() {
        stopProgressTimer()
        progressTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { _ in
            self.updateProgress()
        }
    }
    
    private func stopProgressTimer() {
        progressTimer?.invalidate()
        progressTimer = nil
    }
    
    private func updateProgress() {
        guard let player = audioPlayer, currentState == "playing" else { return }
        
        currentPosition = player.currentTime
        let progress = duration > 0 ? currentPosition / duration : 0.0
        
        let info: [String: Any] = [
            "currentPositionMs": Int(currentPosition * 1000),
            "durationMs": Int(duration * 1000),
            "progress": progress
        ]
        
        DispatchQueue.main.async {
            self.progressEventSink?(info)
        }
    }
}

// MARK: - AVAudioPlayerDelegate
extension PlayMidifilePlugin: AVAudioPlayerDelegate {
    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        updateState("stopped")
        stopProgressTimer()
    }
    
    public func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        updateState("error")
        stopProgressTimer()
    }
}

// MARK: - FlutterStreamHandler
extension PlayMidifilePlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        if let channelName = arguments as? String {
            if channelName == "progress" {
                progressEventSink = events
            } else if channelName == "state" {
                stateEventSink = events
            }
        } else {
            // 通过方法channel判断
            if progressEventSink == nil {
                progressEventSink = events
            } else {
                stateEventSink = events
            }
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        progressEventSink = nil
        stateEventSink = nil
        return nil
    }
} 